### ISSUE TITLE
Terrible use case leading to lose attachments

### ISSUE THREAD

**Issue by e-gaulue** [wekan/wekan#4357]
### Server Setup Information
* Did you test in newest Wekan: yes debian snap install with auto update
* Operating System: Debian buster
* Deployment Method (Snap/Docker/Sandstorm/bundle/source): snap
* Http frontend if any (Caddy, Nginx, Apache, see config examples from Wekan GitHub wiki first): Apache
* Node.js Version: snap is not concerned
* MongoDB Version: the one coming with wekan

### Problem description

Maybe same than #2635. For no reason, while we requested duplication of a board we got hundred of them (partial duplicate). We decided to archive them, and then to delete them. Doing this we lost all images of the initial board.

#### Reproduction Steps

Create a board, add a card, add an attachment to the card, duplicate the board, then archive and delete the duplicated board => the attachment of the initial card in the initial board is lost.

Looks like card duplication don't duplicate the attachments, but just its reference. In this case, board/card deletion shouldn't delete the attachments and another tool should help us to delete those orphaned attachement like the orphaned attachment report. 

#### Logs
Don't know if it's linked but I observe:
```
2022-02-14T10:19:49+01:00 wekan.wekan[29146]: Exception while invoking method 'toggleCardMaximized' TypeError: Cannot read property 'toggleCardMaximized' of null
2022-02-14T12:22:33+01:00 wekan.wekan[29146]: Exception in removedobserveChanges callback: Error: Removed nonexistent document H3vssYg8d4jsmzuXJ
2022-02-14T12:22:33+01:00 wekan.wekan[29146]: Exception in removedobserveChanges callback: Error: Removed nonexistent document H3vssYg8d4jsmzuXJ
2022-02-14T12:22:34+01:00 wekan.wekan[29146]: Exception in removedobserveChanges callback: Error: Removed nonexistent document kjdT4Z2S3L82K9S8D
2022-02-14T12:22:34+01:00 wekan.wekan[29146]: Exception in removedobserveChanges callback: Error: Removed nonexistent document kjdT4Z2S3L82K9S8D
2022-02-14T12:22:34+01:00 wekan.wekan[29146]: Exception in removedobserveChanges callback: Error: Removed nonexistent document kjdT4Z2S3L82K9S8D
2022-02-14T12:22:34+01:00 wekan.wekan[29146]: Exception in removedobserveChanges callback: Error: Removed nonexistent document kjdT4Z2S3L82K9S8D
2022-02-14T12:22:34+01:00 wekan.wekan[29146]: Exception in removedobserveChanges callback: Error: Removed nonexistent document vZL96ekn64gQvmMAq
2022-02-14T12:22:34+01:00 wekan.wekan[29146]: Exception in removedobserveChanges callback: Error: Removed nonexistent document vZL96ekn64gQvmMAq
2022-02-14T12:22:34+01:00 wekan.wekan[29146]: Exception in removedobserveChanges callback: Error: Removed nonexistent document ZCveijGtd8hxv6PWJ
2022-02-14T12:22:34+01:00 wekan.wekan[29146]: Exception in removedobserveChanges callback: Error: Removed nonexistent document ZCveijGtd8hxv6PWJ
```

#### Request

Could anyone help me to bring back those attachments from my mongodb backup? I'm not really confident with MongoDB.

Regards

**Comment by xet7** [comment]
@e-gaulue 

1. Install MongoDB, for example version 4.4 from https://www.mongodb.com/try/download/community
2. Install MongoDB tools, includes mongorestore etc https://www.mongodb.com/try/download/database-tools
3. Install MongoDB GUI nosqlboster https://github.com/wekan/wekan/wiki/Backup#using-nosqlbooster-closed-source-mongodb-gui-with-wekan-snap-to-edit-mongodb-database
4. Run command `mongorestore` in directory where is previous WeKan backup in subdirectory `dump`
5. With nosqlbooster, connect to localhost port 27017 , look at database wekan, at some attachment table there is a list of files. You can click first, shift-click last, save all attachments as files to your harddisk directory.

In future, there will be attachments related fixes, moving attachments from MongoDB database to filesystem or S3, etc.

**Comment by e-gaulue** [comment]
@xet7 

Here is a summary of our last ten days with wekan:
 * last previous sunday evening and previous monday morning two users complain they couldn't add image, instead of it there was just a white square but not the image
 * last sunday (February 13th) we duplicate our main board that lead to hundreds of them looking incomplete
 * we archived all those wrong duplicates
 * I removed one by one all of them and that took time as I had to confirm deletion each time (it was the day before yesterday night)
 * Because of that, images disappeared
 * yesterday night, I tried to restore backup, but before I did anything, images were back, but also the hundred of wrong archive duplicates (just two were really delete and the other were back)

That's a good news, because anyway, my backup process was down for about one year. So I didn't have any.

Because images were still there, I managed to get them following your nosqlbooster steps.

Just to document:
**The reproduction steps I provided in my initial message is right, no trace of the image I put as attachment in my example can be found.**

When we request the duplication of the board we had about as much errors in the log as wrong duplicated board:
```
2022-02-13T20:23:20+01:00 wekan.wekan[4683]: /snap/wekan/1981/programs/server/npm/node_modules/meteor/wekan-cfs-gridfs/node_modules/mongodb/lib/utils.js:98
2022-02-13T20:23:20+01:00 wekan.wekan[4683]:     process.nextTick(function() { throw err; });
2022-02-13T20:23:20+01:00 wekan.wekan[4683]:                                   ^
2022-02-13T20:23:21+01:00 wekan.wekan[4683]: Error [MongoError]: file with id 62095aa8aeecc215c292805f not opened for writing
2022-02-13T20:23:21+01:00 wekan.wekan[4683]:     at Function.MongoError.create (/snap/wekan/1981/programs/server/npm/node_modules/meteor/wekan-cfs-gridfs/node_modules/mongodb-core/lib/error.js:31:11)
2022-02-13T20:23:21+01:00 wekan.wekan[4683]:     at /snap/wekan/1981/programs/server/npm/node_modules/meteor/wekan-cfs-gridfs/node_modules/mongodb/lib/gridfs/grid_store.js:979:33
2022-02-13T20:23:21+01:00 wekan.wekan[4683]:     at handleCallback (/snap/wekan/1981/programs/server/npm/node_modules/meteor/wekan-cfs-gridfs/node_modules/mongodb/lib/utils.js:96:12)
2022-02-13T20:23:21+01:00 wekan.wekan[4683]:     at /snap/wekan/1981/programs/server/npm/node_modules/meteor/wekan-cfs-gridfs/node_modules/mongodb/lib/collection.js:1401:5
2022-02-13T20:23:21+01:00 wekan.wekan[4683]:     at handleCallback (/snap/wekan/1981/programs/server/npm/node_modules/meteor/wekan-cfs-gridfs/node_modules/mongodb/lib/utils.js:96:12)
2022-02-13T20:23:21+01:00 wekan.wekan[4683]:     at /snap/wekan/1981/programs/server/npm/node_modules/meteor/wekan-cfs-gridfs/node_modules/mongodb/lib/cursor.js:675:5
2022-02-13T20:23:21+01:00 wekan.wekan[4683]:     at handleCallback (/snap/wekan/1981/programs/server/npm/node_modules/meteor/wekan-cfs-gridfs/node_modules/mongodb-core/lib/cursor.js:172:5)
2022-02-13T20:23:21+01:00 wekan.wekan[4683]:     at setCursorNotified (/snap/wekan/1981/programs/server/npm/node_modules/meteor/wekan-cfs-gridfs/node_modules/mongodb-core/lib/cursor.js:517:3)
2022-02-13T20:23:21+01:00 wekan.wekan[4683]:     at /snap/wekan/1981/programs/server/npm/node_modules/meteor/wekan-cfs-gridfs/node_modules/mongodb-core/lib/cursor.js:590:16
2022-02-13T20:23:21+01:00 wekan.wekan[4683]:     at queryCallback (/snap/wekan/1981/programs/server/npm/node_modules/meteor/wekan-cfs-gridfs/node_modules/mongodb-core/lib/cursor.js:233:18)
2022-02-13T20:23:21+01:00 wekan.wekan[4683]:     at /snap/wekan/1981/programs/server/npm/node_modules/meteor/wekan-cfs-gridfs/node_modules/mongodb-core/lib/connection/pool.js:436:18
2022-02-13T20:23:21+01:00 wekan.wekan[4683]:     at processTicksAndRejections (internal/process/task_queues.js:79:11)
2022-02-13T20:23:21+01:00 wekan.wekan[4683]: Emitted 'error' event on GridReadStream instance at:
2022-02-13T20:23:21+01:00 wekan.wekan[4683]:     at GridReadStream._error (/snap/wekan/1981/programs/server/npm/node_modules/meteor/wekan-cfs-gridfs/node_modules/gridfs-stream/lib/readstream.js:231:8)
2022-02-13T20:23:21+01:00 wekan.wekan[4683]:     at /snap/wekan/1981/programs/server/npm/node_modules/meteor/wekan-cfs-gridfs/node_modules/gridfs-stream/lib/readstream.js:109:26
2022-02-13T20:23:21+01:00 wekan.wekan[4683]:     at /snap/wekan/1981/programs/server/npm/node_modules/meteor/wekan-cfs-gridfs/node_modules/mongodb/lib/gridfs/grid_store.js:238:22
2022-02-13T20:23:21+01:00 wekan.wekan[4683]:     at error (/snap/wekan/1981/programs/server/npm/node_modules/meteor/wekan-cfs-gridfs/node_modules/mongodb/lib/gridfs/grid_store.js:1061:5)
2022-02-13T20:23:21+01:00 wekan.wekan[4683]:     at /snap/wekan/1981/programs/server/npm/node_modules/meteor/wekan-cfs-gridfs/node_modules/mongodb/lib/gridfs/grid_store.js:979:16
2022-02-13T20:23:21+01:00 wekan.wekan[4683]:     at handleCallback (/snap/wekan/1981/programs/server/npm/node_modules/meteor/wekan-cfs-gridfs/node_modules/mongodb/lib/utils.js:96:12)
2022-02-13T20:23:21+01:00 wekan.wekan[4683]:     [... lines matching original stack trace ...]
2022-02-13T20:23:21+01:00 wekan.wekan[4683]:     at processTicksAndRejections (internal/process/task_queues.js:79:11) {
2022-02-13T20:23:21+01:00 wekan.wekan[4683]:   driver: true
2022-02-13T20:23:21+01:00 wekan.wekan[4683]: }
2022-02-13T20:23:21+01:00 systemd[1]: snap.wekan.wekan.service: Main process exited, code=exited, status=1/FAILURE
2022-02-13T20:23:21+01:00 systemd[1]: snap.wekan.wekan.service: Failed with result 'exit-code'.
2022-02-13T20:23:21+01:00 systemd[1]: snap.wekan.wekan.service: Service RestartSec=100ms expired, scheduling restart.
2022-02-13T20:23:21+01:00 systemd[1]: snap.wekan.wekan.service: Scheduled restart job, restart counter is at 15.
2022-02-13T20:23:21+01:00 systemd[1]: Stopped Service for snap application wekan.wekan.
2022-02-13T20:23:21+01:00 systemd[1]: Started Service for snap application wekan.wekan.
```

So now, I don't know what to do with all those wrong duplicated boards. We have a main board we duplicate every Sunday, to keep an history of what we did. We have about 60 copies in our timelapse now, and we want to keep those as they are. I worry if I remove the wrong duplicated ones, I'll loose some images.

By the way, what are _tempstore.files? Resized attachments/images? Is there a way to force their rebuilding?

Regards,

**Comment by xet7** [comment]
@e-gaulue 

1. You can look at those attachment IDs, do they belong to some card ID, that belongs to specific board ID, etc.
2. When you have found correct boards, rename them, so you know the correct boards.
3. At All Boards page hover mouse over other non-correct board icon, click Archive Board, to archive those board.
4. At Board Archive, delete those boards.

**Comment by xet7** [comment]
@e-gaulue 

Also, use this database query to delete incomplete uploads, if there are any:
https://github.com/wekan/wekan/issues/1298#issuecomment-1031041529

**Comment by e-gaulue** [comment]
Hi @xet7,

Since your last answer, I've been on holidays, reason why I was quiet.

Few remarks about our use case. We have a main board, every week we create cards on or archived them. Then we duplicate the board which become a "screenshot" of the board at this day.

Since about 1 month now (I can find the exact day if needed), our instance has get unusable the way describe above.

About one month ago, we requested to duplicate our main board but while processing we got the message under, leading to wekan to restart and another duplication to occur, new error, new restart... Because of this, we got about a hundreds duplicates of our main board. Why it stopped, I don't know, but it helped us.

```
wekan.wekan[4683]: /snap/wekan/1981/programs/server/npm/node_modules/meteor/wekan-cfs-gridfs/node_modules/mongodb/lib/utils.js:98
wekan.wekan[4683]:     process.nextTick(function() { throw err; });
wekan.wekan[4683]:                                   ^
wekan.wekan[4683]: Error [MongoError]: file with id 62095aa8aeecc215c292805f not opened for writing
``` 

That's where this issue has started. Because of the message we believed the trouble was due to our attachments, and as some have disappeared at this time, I sent those posts you read.

But, each time we have requested a duplication of our main board again, the same story was going on. And because of the massage again, I though it was due to attachment. So I followed this issue #1740: Export the mongo DB to json and grep for the file id.

Doing this we observed, our archived cards where certainly also duplicated while copying the board. In fact we never removed archived cards from our main board, reason why maybe our user found the board slower. So we tried to used the interface to remove "all" archived card with the button at the top from board engine button section "archive". But it didn't work (BUGG???). Looks like it removed just the first or some but not all.

Never mind, I did removed the archived cards this way:
```
db.cards.deleteMany({boardId: 'kZF23soqDqqKWTmxf', archived: true})
```

Then I removed the orphaned cfs_attachments.filrecord this way:
```
db.cfs.attachments.filerecord.aggregate([])
    .lookup({
         from: "cards",
         localField: "cardId",
         foreignField: "_id",
         as: "inventory_docs"
})
    .project({boardId:1, 'countcard': {$size: "$inventory_docs"}})
    .match({'countcard': {$lt: 1}})
    .project({'_id':1})
    .limit(500)
```
As I'm new to MongoDB I didn't manage to remove them from script, so I needed to use "delete rows" button from NoSQLBooster. A little trick to do it would be greatly appreciated.

I did it also for swinlanes... So at the end, I had a clean board, lists, cards, attachement in MongoDB at least for our main board, but when I tried to request duplication, it is still the same:
```
wekan.wekan[4683]: /snap/wekan/1981/programs/server/npm/node_modules/meteor/wekan-cfs-gridfs/node_modules/mongodb/lib/utils.js:98
wekan.wekan[4683]:     process.nextTick(function() { throw err; });
wekan.wekan[4683]:                                   ^
wekan.wekan[4683]: Error [MongoError]: file with id 6225cafdb5c51853e23c6b35 not opened for writing
``` 
So with another key.

But this time, it was not possible to find any reference to this key in the json export of the mongo DB. So we believe the trouble is maybe on the wekan side. We don't know what is going on. When you clic on the duplicate button, you have the feeling nothing happen for few minutes and then wekan crash, restart and... produce multiple duplicates of the original request. Only thing I can says, some attachments are on multiple cards: the ones in our main board but also the ones on our weekly duplicated boards. Can't say if it's the reason of our trouble, but every thing was working well 40 days ago.

I see on github a lot of work is done around fs. Last year, we tried to migrate our CRM with 400k communications to wekan, but it failed because of it. FS request where too long and didn't finish before the other one was starting. Of course, it was on a totally different machine.

I hope this message will give you ideas of what we could do to be able to duplicate our main board again.

Regards, Edouard
